server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/SegmentSlotRelation.java

```diff
@@ -0,0 +1,57 @@
+package org.infinispan.server.resp.commands.cluster;
+
+import static org.infinispan.commons.logging.Log.CONFIG;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.infinispan.distribution.ch.impl.CRC16HashFunctionPartitioner;
+
+public class SegmentSlotRelation {
+
+   static final int SLOT_SIZE = 1 << 14;
+   private final int segmentSize;
+   private final int width;
+   private final Map<Integer, List<Integer>> relation;
+
+   public SegmentSlotRelation(int segmentSize) {
+      if (!CRC16HashFunctionPartitioner.isPow2(segmentSize))
+         throw CONFIG.respCacheSegmentSizePow2(segmentSize);
+      this.segmentSize = segmentSize;
+      this.width = 1 << Integer.numberOfTrailingZeros(SLOT_SIZE) - Integer.numberOfTrailingZeros(segmentSize);
+      Map<Integer, List<Integer>> relation = new HashMap<>(segmentSize);
+      for (int i = 0; i < SLOT_SIZE; i++) {
+         Integer k = i % segmentSize;
```

Comment:
What if we try allocating width slots to each segment sequentially

segment 0 = 0 to SLOT_SIZE/numSegments - 1 slots
segment 1 = SLOT_SIZE/numSegments to (2 * SLOT_SIZE/numSegments) -1
...
segment 255 = (255 * SLOT_SIZE/numSegments) to (256 * SLOT_SIZE/numSegments) -1

This should give much closer groupings of slots, which should make your description with the slots much more concise.

Reply:
I like the idea of making it contiguous so we don't return that big response, but I couldn't make it work :frowning_face:. A concrete example from the tests is a key with `hash=46020`, Lettuce maps to `slot=13252`, and the KeyPartitioner maps to `segment=196`.

If we have this contiguous mapping, segments 196 maps to `[12544, 12608)`, not covering 13252. Unless I am missing something, we would need to change the partitioner.

On the other hand, I could make the current "scattered" mapping between the slots and segments working with the suggestions, getting rid of the Map, and using just math. The downside is the big fat response.

Reply:
Ah, it is because the partitioner already does something different for segments than the others

https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/distribution/ch/impl/CRC16HashFunctionPartitioner.java#L39

Where as the default HashFunctionPartitioner uses contiguous 

https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/distribution/ch/impl/HashFunctionPartitioner.java#L66

Note that segmentSize is Integer.MAX_VALUE / numSegments so it is divided into blocks.

Reply:
Looking closer the other issue is that the CRC16 hash only returns a hash code with up to 16 bits which normally we do 32 bits.

So in this case the hash partitioner would need to be aware of the hash max value. And change the line at https://github.com/infinispan/infinispan/blob/main/commons/all/src/main/java/org/infinispan/commons/util/Util.java#L704 to take the max hashValue instead of Integer.MAX_VALUE to determine the segment size.